### PR TITLE
Only call prepareForDraw when needed (better performance)

### PR DESCRIPTION
**Pull request by pmolodo**
## Description (this won't be part of the changelog)
Currently, the ProxyDrawOverride is set to always be considered "dirty" whenever a redraw is requested, which triggers a "prepareForDraw" call.   This call is the only place from which it is allowed to pull data from the proxyShape, but it is also somewhat slow as a result.   By setting the "isAlwaysDirty" property to false, we can get better performance, particularly when the proxy isn't changing.

The only downside of setting isAlwaysDirty=false is we need to make sure we notify maya when the shape needs to be updated, but this wasn't too hard to do.

I did some performance tests that combined this branch, and PR 89, since it's also performance related:

https://github.com/AnimalLogic/AL_USDMaya/pull/89

My tests showed ~25% speedup in viewport tumble speed (from 1.9 fps to 2.4 fps).  Granted, it was a test environment designed to hit the "prepareForDraw" call as much as possible, but it did show that this is having an impact.  I can provide the test scene I used if you want (basically, it was just 1000 proxyShapes, each drawing a single cube).

### Changed
- Improved performance when proxy shape isn't changing (ie, viewport tumble)

## Checklist (Please do not remove this line)
- [X] Make sure the Title and Description of the PR make sense and  provide sufficient context for your work
- [X] Do any added files have the correct AL Apache Licence Header?
- [X] Are there Doxygen comments in the headers?
- [X] Are any new features, behaviour changes documented in the .md format [documentation](https://github.com/AnimalLogic/AL_USDMaya/docs)?
- [X] Have you added, updated tests to cover new features and behaviour changes?
- [X] Have you filled out at least one changelog entry?

### CODE REVIEW

**Comment by elrond79** [comment]
Re-pushed with some fixes for the way that a changed xform was detected, and the return value of connectionMade/connectionBroken

**Comment by elrond79** [comment]
Re-pushed with some fixes for the way that a changed xform was detected, and the return value of connectionMade/connectionBroken

File: lib/AL_USDMaya/AL/usdmaya/nodes/ProxyDrawOverride.cpp [hunk 1]
```diff
@@ -132,6 +138,7 @@ MUserData* ProxyDrawOverride::prepareForDraw(
 
   if(!data->m_shape || !data->m_shape->getRenderAttris(&data->m_params, frameContext, objPath))
   {
+    if (newData) delete data;
```
Comment (robthebloke) [inline lib/AL_USDMaya/AL/usdmaya/nodes/ProxyDrawOverride.cpp]: This looks to be a little backward? 
Would it be better to see if fn.userNode() points to something valid, then create the data, if not, don't?

Comment (robthebloke) [reply lib/AL_USDMaya/AL/usdmaya/nodes/ProxyDrawOverride.cpp]: If that doesn't remove the need for the newData bool, then I guess something along the lines of:

```c++
RenderUserData* newData = nullptr;
if(!data)
{
  data = newData = new RenderUserData;
}

// .... then later, just.... 
delete newData;

```

**Review by robthebloke — COMMENTED** [review]
Mostly looks good, but i feel a small tidy up of some of the logic is possible in the noted area....

**Comment by elrond79** [comment]
So, I reorganized the code in ProxyDrawOverride::prepareForDraw() so that all checks happen first, and all assignments happen at the end.  We ensure that the data object is fully populated before returning it, which also avoids a (potential) crash in the original code, if shape->engine() returned a nullptr.  (Because the RenderUserData* was still returned, but when used by draw later, it did not check if m_engine was null).

However, there is one more optimization that could be done - removing the check of getRenderAttris()'s return value altogether. I say this because there is no code path where it returns false. I'd actually like to change the signature for it to return void, so we can remove this check (I feel like as long as the signature is to return a bool, presumably for success, the "right" thing to do is to check it, even if it currently always returns true).

Alternatively, we could make getRenderAttris() actually return False - ie, there are some maya api calls whose MStatus is not checked, and we could have it return false if those fail.

**Review by robthebloke — APPROVED** [review]

**Comment by aloysbaillet** [comment]
Hi @elrond79 ,

We just noticed that this changes breaks animated stage playback when in parallel evaluation (works fine in DG while the time node is connected).
As we're a bit short on staff at the moment we are just doing a quick fix by changing the default to isAlwaysDirty back to true...
Do you think you could have a look at making this change work in parallel evaluation as well?

Cheers,

Aloys

**Comment by chadrik** [comment]
Hi Aloys, Paul is on holiday for 2 weeks.  What's the priority level on
your end?  I can have someone else take a look if you feel it's pressing.

Also, we should probably get a test in place for this and create an issue
for running tests in both parallel and DG modes.

**Comment by aloysbaillet** [comment]
Hi Chad! Don't worry about it for now, we just reverted the default value for this optimisation so we are good. Would be great to revisit with Paul as he knew what he was trying to optimise exactly, but that can definitely wait for his return.

**Comment by fleith** [comment]
Just complementing, here at Autodesk, we found problems on two operations, duplicating objects (Ctrl + D), and deleting then undo. After a regression search, we found that the commit https://github.com/AnimalLogic/AL_USDMaya/commit/321be29175a43aa270bb9d50387ac335c0d483d0 is causing the crashes. Now with the flag isAllwaysDirty true, works fine.

**Pull request merged**